timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/LockWatchValueIntegrationTest.java

```diff
@@ -0,0 +1,280 @@
+            assertLoadedValues(txn, ImmutableMap.of());
+            return values;
+        });
+
+        assertThat(result).containsEntry(CELL_1, DATA_1);
+        assertThat(result).containsExactlyInAnyOrderEntriesOf(result2);
+    }
+
+    @Test
+    public void readWriteTransactionsPublishValuesToCentralCache() {
+        putValue();
+
+        Map<Cell, byte[]> result = txnManager.runTaskThrowOnConflict(txn -> {
+            txn.put(TABLE_REF, ImmutableMap.of(CELL_2, DATA_2));
+            Map<Cell, byte[]> values = txn.get(TABLE_REF, ImmutableSet.of(CELL_1, CELL_2));
+            assertHitValues(txn, ImmutableSet.of());
+            assertLoadedValues(txn, ImmutableMap.of(TABLE_CELL_1, CacheValue.of(DATA_1)));
+            return values;
+        });
+
+        awaitUnlock();
+
+        Map<Cell, byte[]> result2 = txnManager.runTaskThrowOnConflict(txn -> {
+            Map<Cell, byte[]> values = txn.get(TABLE_REF, ImmutableSet.of(CELL_1, CELL_2));
+            assertHitValues(txn, ImmutableSet.of(TABLE_CELL_1));
+            assertLoadedValues(txn, ImmutableMap.of(TABLE_CELL_2, CacheValue.of(DATA_2)));
+            return values;
+        });
+
+        assertThat(result).containsEntry(CELL_1, DATA_1);
+        assertThat(result).containsEntry(CELL_2, DATA_2);
+        assertThat(result).containsExactlyInAnyOrderEntriesOf(result2);
+    }
+
+    @Test
+    public void serializableTransactionsThrowOnCachedReadWriteConflicts() {
+        putValue();
+
+        // Nested transactions are normally discouraged, but it is easier to create a conflict this way (compared to
+        // using executors, which is much more heavy-handed).
+        assertThatThrownBy(() -> txnManager.runTaskThrowOnConflict(txn -> {
+                    txn.get(TABLE_REF, ImmutableSet.of(CELL_1));
+                    txn.put(TABLE_REF, ImmutableMap.of(CELL_2, DATA_2));
+
+                    txnManager.runTaskThrowOnConflict(txn2 -> {
+                        txn2.put(TABLE_REF, ImmutableMap.of(CELL_1, DATA_3));
+                        return null;
+                    });
+                    awaitUnlock();
+                    return null;
+                }))
+                .isExactlyInstanceOf(TransactionSerializableConflictException.class)
+                .hasMessageContaining("There was a read-write conflict on table");
+    }
+
+    @Test
+    public void serializableTransactionsDoNotThrowWhenOverwritingAPreviouslyCachedValue() {
+        putValue();
+
+        assertThatCode(() -> txnManager.runTaskThrowOnConflict(txn -> {
+                    txn.get(TABLE_REF, ImmutableSet.of(CELL_1));
+                    txn.put(TABLE_REF, ImmutableMap.of(CELL_1, DATA_3));
+                    return null;
+                }))
+                .doesNotThrowAnyException();
+    }
+
+    @Test
+    public void serializableTransactionsReadViaTheCacheWherePossible() {
+        putValue();
+
+        txnManager.runTaskThrowOnConflict(txn -> {
```

Comment:
Maybe add a test that we actually invalidate a cached entry once we write to it for the subsequent transactions?